Source/Core/Core/AchievementManager.cpp

```diff
@@ -146,80 +152,208 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
           },
       .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
   rc_hash_init_custom_filereader(&volume_reader);
-  if (!rc_hash_generate_from_file(m_game_hash.data(), RC_CONSOLE_GAMECUBE, iso_path.c_str()))
+  m_queue.EmplaceItem([this, callback, file_path] {
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, file_path.c_str()))
+    {
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file {}.",
+                    file_path);
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
+    }
+    {
+      std::lock_guard lg{m_lock};
+      m_game_hash = std::move(new_hash);
+    }
+    LoadGameSync(callback);
+  });
+}
+
+void AchievementManager::HashGame(const DiscIO::Volume* volume, const ResponseCallback& callback)
+{
+  if (!m_is_runtime_initialized)
+  {
+    ERROR_LOG_FMT(ACHIEVEMENTS,
+                  "Attempted to load game achievements without Achievement Manager initialized.");
+    callback(AchievementManager::ResponseType::MANAGER_NOT_INITIALIZED);
+    return;
+  }
+  if (m_disabled)
   {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from game file.");
+    INFO_LOG_FMT(ACHIEVEMENTS, "Achievement Manager is disabled until core is rebooted.");
+    OSD::AddMessage("Achievements are disabled until core is rebooted.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
     return;
   }
+  m_loading_volume = DiscIO::CreateVolume(volume->GetBlobReader().CopyReader());
```

Comment:
Let's say `HashGame` gets called twice in a row, with very little time in between. What's stopping the second invocation of `HashGame` from overwriting `m_loading_volume` before the first enqueued action moves from `m_loading_volume`?